### ISSUE TITLE
problems with large sets, lists...

### ISSUE THREAD

**Issue by merl-dev** [jkaye2012/Redis.jl#1]
Really nice library and we have been using it in production to update our redis servers, but it's got some issues when retrieving lots of data.  Additionally, when errors do occur, the TcpSocket buffer isn't cleared so the next request fails.  There are no issues we have seen when making small requests so our solution thus far has been to limit response size to at most 100 items. 

Consider the following
julia> for i=1:10000
       lpush(conn,"test_list",string("list_item:",i))
       end

julia> lrange(conn,"test_list",0,-1)
ERROR: ProtocolException("\$14")
 in parse_array_reply at /home/ubuntu/.julia/v0.3/Redis/src/parser.jl:77
 in parse_reply at /home/ubuntu/.julia/v0.3/Redis/src/parser.jl:18
 in execute_command at /home/ubuntu/.julia/v0.3/Redis/src/connection.jl:104
 in lrange at /home/ubuntu/.julia/v0.3/Redis/src/client.jl:68

julia> lrange(conn,"test_list",0,-1)
ERROR: ProtocolException("\r\nlist_item:9380\r\n\$14\r\nlist_item:9379\r\n\$14\r\nlist_item:9378\r\n\$14\r\nlist_item:9377\r\n\$14\r\nlist_item:9376\r\n\$14\r\nlist_item:9375\r\n\$14\r\nlist_item:9374\r\n\$14\r\nlist_item:9373\r\n\$14\r\nlist_item:9372\r\n\$14\r\nlist_item:9371\r\n\$14\r\nlist_item:9370\r\n\$14\r\nlist_item:9369\r\n\$14\r\nlist_item:9368\r\n\$14\r\nlist_item:9367\r\n\$14\r\nlist_item:9366\r\n\$14\r\nlist_item:9365\r\n\$14\r\nlist_item:9364\r\n\$14\r\nlist_item:9363\r\n\$14\r\nlist_item:9362\r\n\$14\r\nlist_item:9361\r\n\$14\r\nlist_item:9360\r\n\$14\r\nlist_item:9359\r\n\$14\r\nlist_item:9358\r\n\$14\r\nlist_item:9357\r\n\$14\r\nlist_item:9356\r\n\$14\r\nlist_item:9355\r\n\$14\r\nlist_item:9354\r\n\$14\r\nlist_item:9353\r\n\$14\r\nlist_item:9352\r\n\$14\r\nlist_item:9351\r\n\$14\r\nlist_item:9350\r\n\$14\r\nlist_item:9349\r\n\$14\r\nlist_item:9348\r\n\$14\r\nlist_item:9347\r\n\$14\r\nlist_item:9346\r\n\$14\r\nlist_item:9345\r\n\$14\r\nlist_item:9344\r\n\$14\r\nlist_item:9343\r\n\$14\r\nlist_item:9342\r\n\$14\r\nlist_item:9341\r\n\$14\r\nlist_item:9340\r\n\$14\r\nlist_item:9339\r\n\$14\r\nlist_item:9338\r\n\$14\r\nlist_item:9337\r\n\$14\r\nlist_item:9336
.................
in execute_command at /home/ubuntu/.julia/v0.3/Redis/src/connection.jl:104
 in lrange at /home/ubuntu/.julia/v0.3/Redis/src/client.jl:68

julia> lrange(conn,"test_list",0,-1)
10000-element Array{Any,1}:
 "list_item:10000"
 "list_item:9999" 
 "list_item:9998" 
 "list_item:9997" 
 "list_item:9996" 
 "list_item:9995" 
 "list_item:9994" 
 "list_item:9993" 
 "list_item:9992" 
 "list_item:9991" 
 "list_item:9990" 
 "list_item:9989" 
 "list_item:9988" 
 "list_item:9987" 
 "list_item:9986" 
 "list_item:9985" 
 ⋮                
 "list_item:16"   
 "list_item:15"   
 "list_item:14"   
 "list_item:13"   
 "list_item:12"   
 "list_item:11"   
 "list_item:10"   
 "list_item:9"    
 "list_item:8"    
 "list_item:7"    
 "list_item:6"    
 "list_item:5"    
 "list_item:4"    
 "list_item:3"    
 "list_item:2"    
 "list_item:1" 

Same happens with just about any of the redis objects.

**Comment by jkaye2012** [comment]
Hi, glad to hear that you're getting some good use out of the library! From the error, this seems like an oversight in the parsing module. I only tested the library with relatively small data sets. I'll try to take a look into this issue sometime this week.

--Jordan

**Comment by merl-dev** [comment]
great thanks!  

John

**Comment by jkaye2012** [comment]
Sorry this took so long!

This is an issue with the way that [connection.jl](https://github.com/jkaye2012/Redis.jl/blob/master/src/connection.jl#L104) receives data from Redis (or possibly an issue with Julia's TCP library, I am not sure which).

The `readavailable` call is returning incomplete data to the calling code. The remaining data that you're seeing in the TcpSocket buffer is the remainder of the reply that `readavailable` did not supply during the first call. It appears to be nondeterministic how much of the reply the first call will provide. The odd thing is: if you repeat the command many times, it begins to succeed and then continues to succeed for every call after. I will continue to look into this, unless you have some insight into a solution?

This problem has actually exposed another [issue](https://github.com/jkaye2012/Redis.jl/issues/5), which is that I don't provide a way for TcpSocket timeouts to be set through the API.

**Comment by merl-dev** [comment]
Hi...didn't see the update you made, so sorry for the late reply.

I am (or was) at a loss to figure this out.  Tried multiple solutions, and then gave up.  Since we are very heavy Redis users in our production environment, we decided to build a new Redis julia client which combines your approach with the wrapper found in HiRedis.jl.  Turned out really nice and haven't made it public yet, still working on the transaction part and the documentation.  Should be posted in the next month or so.

**Comment by sbromberger** [comment]
This may be related:

```
julia> command(c)
ERROR: Redis.ProtocolException("*6\r\n\$9\r\nrpoplpush\r\n:3\r\n")
 in parse_array_reply at /Users/seth/dev/julia/wip/Redis.jl/src/parser.jl:79
 in parse_reply at /Users/seth/dev/julia/wip/Redis.jl/src/parser.jl:18
 in execute_command at /Users/seth/dev/julia/wip/Redis.jl/src/connection.jl:104
 in command at /Users/seth/dev/julia/wip/Redis.jl/src/client.jl:78
```

The following command should be returning `"bar"` but doesn't:

```
julia> get(c,"foo")
2-element Array{Any,1}:
 "write"
 "denyoom"
```

**Comment by sbromberger** [comment]
I think I fixed this in https://github.com/sbromberger/Redis.jl/tree/newparser - would you mind taking a look?

**Comment by sbromberger** [comment]
This is fixed by #9 for all my tests. Could you confirm?

**Comment by jkaye2012** [comment]
Confirmed fixed.
